File: bindings/kotlin/whatwg/src/test/kotlin/com/whatwg/PlatformTests.kt
package com.whatwg

import org.junit.Test
import org.junit.Assert.*

/**
 * Unit tests for WhatWGPlatform initialization and configuration.
 * 
 * Note: Tests that require native library loading are tested in 
 * instrumented tests (androidTest).
 */
class PlatformTests {

    // MARK: - Platform Creation

    @Test
    fun testPlatformCreation() {
        val platform = WhatWGPlatform()
        assertNotNull(platform)
    }

    @Test
    fun testPlatformWithCustomEngine() {
        val platform = WhatWGPlatform()
        platform.engine = JSEngine.QUICKJS
        assertEquals(JSEngine.QUICKJS, platform.engine)
    }

    @Test
    fun testDefaultEngine() {
        val platform = WhatWGPlatform()
        assertEquals(JSEngine.QUICKJS, platform.engine)
    }

    // MARK: - Capability Configuration

    @Test
    fun testCapabilityEnumValues() {
        assertEquals(0, Capability.CLIPBOARD.value)
        assertEquals(1, Capability.TIMER.value)
        assertEquals(2, Capability.NETWORK.value)
        assertEquals(3, Capability.STORAGE.value)
        assertEquals(4, Capability.LAYOUT.value)
        assertEquals(5, Capability.UI.value)
    }

    @Test
    fun testAllCapabilitiesCount() {
        assertEquals(31, Capability.entries.size)
    }

    @Test
    fun testCapabilityNames() {
        assertEquals("CLIPBOARD", Capability.CLIPBOARD.name)
        assertEquals("NETWORK", Capability.NETWORK.name)
        assertEquals("STORAGE", Capability.STORAGE.name)
        assertEquals("GEOLOCATION", Capability.GEOLOCATION.name)
    }

    // MARK: - Provider Configuration (without native calls)

    @Test
    fun testSetClipboardProviderNull() {
        val platform = WhatWGPlatform()
        platform.clipboardProvider = null
        assertNull(platform.clipboardProvider)
    }

    @Test
    fun testSetTimerProviderNull() {
        val platform = WhatWGPlatform()
        platform.timerProvider = null
        assertNull(platform.timerProvider)
    }

    @Test
    fun testSetNetworkProviderNull() {
        val platform = WhatWGPlatform()
        platform.networkProvider = null
        assertNull(platform.networkProvider)
    }

    @Test
    fun testSetStorageProviderNull() {
        val platform = WhatWGPlatform()
        platform.storageProvider = null
        assertNull(platform.storageProvider)
    }

    // MARK: - JS Engine Tests

    @Test
    fun testJSEngineRecommended() {
        assertEquals(JSEngine.QUICKJS, JSEngine.recommended)
    }

    @Test
    fun testJSEngineDisplayNames() {
        assertEquals("V8", JSEngine.V8.displayName)
        assertEquals("JavaScriptCore", JSEngine.JAVASCRIPT_CORE.displayName)
        assertEquals("QuickJS", JSEngine.QUICKJS.displayName)
    }

    @Test
    fun testJSEngineValues() {
        assertEquals("v8", JSEngine.V8.value)
        assertEquals("jsc", JSEngine.JAVASCRIPT_CORE.value)
        assertEquals("quickjs", JSEngine.QUICKJS.value)
    }

    @Test
    fun testJSEngineIsAvailable() {
        // QuickJS should always be available
        assertTrue(JSEngine.QUICKJS.isAvailable)
    }

    // MARK: - Exception Tests

    @Test
    fun testWhatWGException() {
        val exception = WhatWGException("Test error")
        assertEquals("Test error", exception.message)
    }

    @Test
    fun testWhatWGExceptionWithCause() {
        val cause = RuntimeException("Cause")
        val exception = WhatWGException("Test error", cause)
        assertEquals("Test error", exception.message)
        assertEquals(cause, exception.cause)
    }

    // MARK: - AutoCloseable Tests

    @Test
    fun testPlatformClose() {
        val platform = WhatWGPlatform()
        // Should not throw
        platform.close()
    }

    @Test
    fun testPlatformUseBlock() {
        WhatWGPlatform().use { platform ->
            assertNotNull(platform)
        }
        // Platform should be closed after use block
    }
}

File: bindings/swift/Tests/WhatWGTests/PlatformTests.swift
import XCTest
@testable import WhatWG

/// Tests for WhatWGPlatform initialization and configuration.
final class PlatformTests: XCTestCase {
    
    // MARK: - Initialization
    
    func testPlatformCreation() {
        let platform = WhatWGPlatform()
        XCTAssertNotNil(platform)
    }
    
    func testPlatformWithCustomEngine() {
        let platform = WhatWGPlatform()
        platform.engine = .quickJS
        XCTAssertEqual(platform.engine, .quickJS)
    }
    
    func testDefaultEngine() {
        let platform = WhatWGPlatform()
        // Default should be JSC on Apple platforms
        #if os(iOS) || os(macOS)
        XCTAssertEqual(platform.engine, .javaScriptCore)
        #endif
    }
    
    // MARK: - Version Info
    
    func testVersionString() {
        let version = WhatWGPlatform.version
        XCTAssertFalse(version.isEmpty, "Version string should not be empty")
    }
    
    func testABIVersion() {
        let version = WhatWGPlatform.abiVersion
        XCTAssertGreaterThan(version, 0, "ABI version should be positive")
    }
    
    // MARK: - Capability Configuration
    
    func testCapabilityEnumValues() {
        XCTAssertEqual(Capability.clipboard.rawValue, 0)
        XCTAssertEqual(Capability.timer.rawValue, 1)
        XCTAssertEqual(Capability.network.rawValue, 2)
        XCTAssertEqual(Capability.storage.rawValue, 3)
        XCTAssertEqual(Capability.layout.rawValue, 4)
        XCTAssertEqual(Capability.ui.rawValue, 5)
    }
    
    func testAllCapabilitiesCount() {
        XCTAssertEqual(Capability.allCases.count, 31)
    }
    
    func testCapabilityDisplayNames() {
        XCTAssertEqual(Capability.clipboard.displayName, "Clipboard")
        XCTAssertEqual(Capability.network.displayName, "Network")
        XCTAssertEqual(Capability.storage.displayName, "Storage")
        XCTAssertEqual(Capability.geolocation.displayName, "Geolocation")
    }
    
    // MARK: - Provider Configuration
    
    #if os(iOS) || os(macOS)
    func testSetClipboardProvider() {
        let platform = WhatWGPlatform()
        let provider = iOSClipboardProvider()
        platform.clipboardProvider = provider
        XCTAssertNotNil(platform.clipboardProvider)
    }
    
    func testSetTimerProvider() {
        let platform = WhatWGPlatform()
        let provider = iOSTimerProvider()
        platform.timerProvider = provider
        XCTAssertNotNil(platform.timerProvider)
    }
    
    func testSetNetworkProvider() {
        let platform = WhatWGPlatform()
        let provider = iOSNetworkProvider()
        platform.networkProvider = provider
        XCTAssertNotNil(platform.networkProvider)
    }
    
    func testSetStorageProvider() {
        let platform = WhatWGPlatform()
        let provider = iOSStorageProvider()
        platform.storageProvider = provider
        XCTAssertNotNil(platform.storageProvider)
    }
    
    func testSetMultipleProviders() {
        let platform = WhatWGPlatform()
        platform.clipboardProvider = iOSClipboardProvider()
        platform.timerProvider = iOSTimerProvider()
        platform.networkProvider = iOSNetworkProvider()
        platform.storageProvider = iOSStorageProvider()
        
        XCTAssertNotNil(platform.clipboardProvider)
        XCTAssertNotNil(platform.timerProvider)
        XCTAssertNotNil(platform.networkProvider)
        XCTAssertNotNil(platform.storageProvider)
    }
    #endif
    
    // MARK: - JS Engine Tests
    
    func testJSEngineRecommended() {
        let recommended = JSEngine.recommended
        #if os(iOS) || os(macOS) || os(tvOS) || os(watchOS)
        XCTAssertEqual(recommended, .javaScriptCore)
        #else
        XCTAssertEqual(recommended, .quickJS)
        #endif
    }
    
    func testJSEngineDisplayNames() {
        XCTAssertEqual(JSEngine.javaScriptCore.displayName, "JavaScriptCore")
        XCTAssertEqual(JSEngine.v8.displayName, "V8")
        XCTAssertEqual(JSEngine.quickJS.displayName, "QuickJS")
    }
    
    func testJSEngineAvailability() {
        #if os(iOS) || os(macOS) || os(tvOS) || os(watchOS)
        XCTAssertTrue(JSEngine.javaScriptCore.isAvailable)
        #endif
        XCTAssertTrue(JSEngine.quickJS.isAvailable)
    }
    
    // MARK: - Error Tests
    
    func testErrorDescriptions() {
        XCTAssertNotNil(WhatWGError.initializationFailed.errorDescription)
        XCTAssertNotNil(WhatWGError.contextCreationFailed.errorDescription)
        XCTAssertNotNil(WhatWGError.notInitialized.errorDescription)
        XCTAssertNotNil(WhatWGError.capabilityNotAvailable(.clipboard).errorDescription)
        XCTAssertNotNil(WhatWGError.operationFailed("test").errorDescription)
    }
    
    func testErrorEquality() {
        XCTAssertEqual(
            WhatWGError.capabilityNotAvailable(.clipboard),
            WhatWGError.capabilityNotAvailable(.clipboard)
        )
        XCTAssertNotEqual(
            WhatWGError.capabilityNotAvailable(.clipboard),
            WhatWGError.capabilityNotAvailable(.network)
        )
    }
}
